tests/test_pipeline.py

```diff
@@ -0,0 +1,446 @@
+        return torch.sum(manual_penalty_coeff[strict_features] * manual_multiplier[strict_features] * violation)
+
+    def _manual_simultaneous_roll(self, manual_x, manual_multiplier, features, strict_features):
+        manual_primal_lagrangian = self._manual_primal_lagrangian(manual_x, manual_multiplier, features)
+        manual_observed_multipliers = manual_multiplier.clone()
+        manual_dual_lagrangian = self._manual_dual_lagrangian(manual_x, manual_multiplier, strict_features)
+        manual_x, manual_multiplier = self._manual_primal_step(
+            manual_x, manual_multiplier, features
+        ), self._manual_dual_step(manual_x, manual_multiplier, strict_features)
+        return (
+            manual_x,
+            manual_multiplier,
+            manual_primal_lagrangian,
+            manual_dual_lagrangian,
+            manual_observed_multipliers,
+        )
+
+    def _manual_dual_primal_roll(self, manual_x, manual_multiplier, features, strict_features, manual_penalty_coeff):
+        manual_dual_lagrangian = self._manual_dual_lagrangian(
+            manual_x, manual_multiplier, strict_features, manual_penalty_coeff
+        )
+        manual_multiplier = self._manual_dual_step(manual_x, manual_multiplier, strict_features, manual_penalty_coeff)
+        manual_primal_lagrangian = self._manual_primal_lagrangian(
+            manual_x, manual_multiplier, features, manual_penalty_coeff
+        )
+        manual_observed_multipliers = manual_multiplier.clone()
+        manual_x = self._manual_primal_step(manual_x, manual_multiplier, features, manual_penalty_coeff)
+        return (
+            manual_x,
+            manual_multiplier,
+            manual_primal_lagrangian,
+            manual_dual_lagrangian,
+            manual_observed_multipliers,
+        )
+
+    def _manual_primal_dual_roll(self, manual_x, manual_multiplier, features, strict_features, manual_penalty_coeff):
+        manual_primal_lagrangian = self._manual_primal_lagrangian(
+            manual_x, manual_multiplier, features, manual_penalty_coeff
+        )
+        manual_observed_multipliers = manual_multiplier.clone()
+        manual_x = self._manual_primal_step(manual_x, manual_multiplier, features, manual_penalty_coeff)
+        manual_dual_lagrangian = self._manual_dual_lagrangian(
+            manual_x, manual_multiplier, strict_features, manual_penalty_coeff
+        )
+        manual_multiplier = self._manual_dual_step(manual_x, manual_multiplier, strict_features, manual_penalty_coeff)
+        return (
+            manual_x,
+            manual_multiplier,
+            manual_primal_lagrangian,
+            manual_dual_lagrangian,
+            manual_observed_multipliers,
+        )
+
+    def _manual_extragradient_roll(self, manual_x, manual_multiplier, features, strict_features):
+        manual_x_copy = manual_x.clone()
+        manual_multiplier_copy = manual_multiplier.clone()
```

Comment:
Would be worth making it clear which variables hold the t+1/2 extrapolated points, and which hold the previous iterates at t.